src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -648,38 +648,58 @@ protected override bool TryDequeue(Task task)
         private void RunOnce(bool checkConditions)
         {
             // Run as long as we have scheduled tasks
-            // TODO(cretz): Fix to run as long as any tasks not yielded on Temporal
             while (scheduledTasks.Count > 0)
             {
-                while (scheduledTasks.Count > 0)
+                // Run all tasks until empty
+                RunAllTasks();
+
+                // If there are any conditions, schedule a new condition-check task and the run all
+                // tasks again. This is a scheduled task instead of just run inline because it needs
+                // to be in the workflow context (i.e. current task scheduler) so the `Workflow`
+                // methods work properly.
+                if (checkConditions && conditions.Count > 0)
                 {
-                    // Pop last
-                    var task = scheduledTasks.Last!.Value;
-                    scheduledTasks.RemoveLast();
-                    scheduledTaskNodes.Remove(task);
+                    _ = QueueNewTaskAsync(CheckConditionsAsync);
```

Comment:
For my own understanding in dotnet, Do we detect if the user tries to call any illegal APIs in the wait condition?

Reply:
We don't and we probably should. We have been mostly counting on the fact that it's not async to prevent commands, but there are commands like upsert search attribute that are not async. This is actually a gap for queries and update validators too. I have opened #243.